File: core/java/android/window/flags/DesktopModeFlags.java
/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.window.flags;

import android.annotation.Nullable;
import android.app.ActivityThread;
import android.app.Application;
import android.content.ContentResolver;
import android.provider.Settings;
import android.util.Log;

import com.android.window.flags.Flags;

import java.util.function.Supplier;

/**
 * Checks desktop mode flag state.
 *
 * <p>This enum provides a centralized way to control the behavior of flags related to desktop
 * windowing features which are aiming for developer preview before their release. It allows
 * developer option to override the default behavior of these flags.
 *
 * <p>NOTE: Flags should only be added to this enum when they have received Product and UX
 * alignment that the feature is ready for developer preview, otherwise just do a flag check.
 *
 * @hide
 */
public enum DesktopModeFlags {
    // All desktop mode related flags to be overridden by developer option toggle will be added here
    ENABLE_DESKTOP_WINDOWING_MODE(
            Flags::enableDesktopWindowingMode, /* shouldOverrideByDevOption= */ true),
    ENABLE_WINDOWING_DYNAMIC_INITIAL_BOUNDS(Flags::enableWindowingDynamicInitialBounds, true),
    ENABLE_CAPTION_COMPAT_INSET_FORCE_CONSUMPTION(
            Flags::enableCaptionCompatInsetForceConsumption, true),
    ENABLE_CAPTION_COMPAT_INSET_FORCE_CONSUMPTION_ALWAYS(
            Flags::enableCaptionCompatInsetForceConsumptionAlways, true),
    ENABLE_CASCADING_WINDOWS(Flags::enableCascadingWindows, true),
    ENABLE_DESKTOP_WINDOWING_WALLPAPER_ACTIVITY(
            Flags::enableDesktopWindowingWallpaperActivity, true),
    ENABLE_DESKTOP_WINDOWING_MODALS_POLICY(Flags::enableDesktopWindowingModalsPolicy, true),
    ENABLE_THEMED_APP_HEADERS(Flags::enableThemedAppHeaders, true),
    ENABLE_DESKTOP_WINDOWING_QUICK_SWITCH(Flags::enableDesktopWindowingQuickSwitch, true),
    ENABLE_APP_HEADER_WITH_TASK_DENSITY(Flags::enableAppHeaderWithTaskDensity, true),
    ENABLE_TASK_STACK_OBSERVER_IN_SHELL(Flags::enableTaskStackObserverInShell, true),
    ENABLE_DESKTOP_WINDOWING_SIZE_CONSTRAINTS(Flags::enableDesktopWindowingSizeConstraints, true),
    DISABLE_NON_RESIZABLE_APP_SNAP_RESIZE(Flags::disableNonResizableAppSnapResizing, true),
    ENABLE_WINDOWING_SCALED_RESIZING(Flags::enableWindowingScaledResizing, true),
    ENABLE_DESKTOP_WINDOWING_TASK_LIMIT(Flags::enableDesktopWindowingTaskLimit, true),
    ENABLE_DESKTOP_WINDOWING_BACK_NAVIGATION(Flags::enableDesktopWindowingBackNavigation, true),
    ENABLE_WINDOWING_EDGE_DRAG_RESIZE(Flags::enableWindowingEdgeDragResize, true),
    ENABLE_DESKTOP_WINDOWING_TASKBAR_RUNNING_APPS(
            Flags::enableDesktopWindowingTaskbarRunningApps, true),
    ENABLE_DESKTOP_WINDOWING_TRANSITIONS(Flags::enableDesktopWindowingTransitions, false);

    private static final String TAG = "DesktopModeFlagsUtil";
    // Function called to obtain aconfig flag value.
    private final Supplier<Boolean> mFlagFunction;
    // Whether the flag state should be affected by developer option.
    private final boolean mShouldOverrideByDevOption;

    // Local cache for toggle override, which is initialized once on its first access. It needs to
    // be refreshed only on reboots as overridden state is expected to take effect on reboots.
    private static ToggleOverride sCachedToggleOverride;

    DesktopModeFlags(Supplier<Boolean> flagFunction, boolean shouldOverrideByDevOption) {
        this.mFlagFunction = flagFunction;
        this.mShouldOverrideByDevOption = shouldOverrideByDevOption;
    }

    /**
     * Determines state of flag based on the actual flag and desktop mode developer option
     * overrides.
     */
    public boolean isTrue() {
        Application application = ActivityThread.currentApplication();
        if (!Flags.showDesktopWindowingDevOption()
                || !mShouldOverrideByDevOption
                || application == null) {
            return mFlagFunction.get();
        } else {
            boolean shouldToggleBeEnabledByDefault = Flags.enableDesktopWindowingMode();
            return switch (getToggleOverride(application.getContentResolver())) {
                case OVERRIDE_UNSET -> mFlagFunction.get();
                // When toggle override matches its default state, don't override flags. This
                // helps users reset their feature overrides.
                case OVERRIDE_OFF -> !shouldToggleBeEnabledByDefault && mFlagFunction.get();
                case OVERRIDE_ON -> shouldToggleBeEnabledByDefault ? mFlagFunction.get() : true;
            };
        }
    }

    private ToggleOverride getToggleOverride(ContentResolver contentResolver) {
        // If cached, return it
        if (sCachedToggleOverride != null) {
            return sCachedToggleOverride;
        }

        // Otherwise, fetch and cache it
        ToggleOverride override = getToggleOverrideFromSystem(contentResolver);
        sCachedToggleOverride = override;
        Log.d(TAG, "Toggle override initialized to: " + override);
        return override;
    }

    /**
     *  Returns {@link ToggleOverride} from Settings.Global set by toggle.
     */
    private ToggleOverride getToggleOverrideFromSystem(ContentResolver contentResolver) {
        int settingValue = Settings.Global.getInt(
                contentResolver,
                Settings.Global.DEVELOPMENT_OVERRIDE_DESKTOP_MODE_FEATURES,
                ToggleOverride.OVERRIDE_UNSET.getSetting()
        );
        return ToggleOverride.fromSetting(settingValue, ToggleOverride.OVERRIDE_UNSET);
    }

    /** Override state of desktop mode developer option toggle. */
    public enum ToggleOverride {
        OVERRIDE_UNSET,
        OVERRIDE_OFF,
        OVERRIDE_ON;

        /** Returns the integer representation of this {@code ToggleOverride}. */
        public int getSetting() {
            return switch (this) {
                case OVERRIDE_ON -> 1;
                case OVERRIDE_OFF -> 0;
                case OVERRIDE_UNSET -> -1;
            };
        }

        /** Returns the {@code ToggleOverride} corresponding to a given integer setting. */
        public static ToggleOverride fromSetting(int setting, @Nullable ToggleOverride fallback) {
            return switch (setting) {
                case 1 -> OVERRIDE_ON;
                case 0 -> OVERRIDE_OFF;
                case -1 -> OVERRIDE_UNSET;
                default -> fallback;
            };
        }
    }
}

File: libs/WindowManager/Shell/shared/src/com/android/wm/shell/shared/desktopmode/DesktopModeFlags.kt
/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.wm.shell.shared.desktopmode

import android.content.Context
import android.provider.Settings
import android.util.Log
import com.android.window.flags.Flags

/*
 * An enum to check desktop mode flags state.
 *
 * This enum provides a centralized way to control the behavior of flags related to desktop
 * windowing features which are aiming for developer preview before their release. It allows
 * developer option to override the default behavior of these flags.
 *
 * NOTE: Flags should only be added to this enum when they have received Product and UX
 * alignment that the feature is ready for developer preview, otherwise just do a flag check.
 */
enum class DesktopModeFlags(
    // Function called to obtain aconfig flag value.
    private val flagFunction: () -> Boolean,
    // Whether the flag state should be affected by developer option.
    private val shouldOverrideByDevOption: Boolean
) {
    // All desktop mode related flags will be added here
    DESKTOP_WINDOWING_MODE(Flags::enableDesktopWindowingMode, true),
    CASCADING_WINDOWS(Flags::enableCascadingWindows, true),
    WALLPAPER_ACTIVITY(Flags::enableDesktopWindowingWallpaperActivity, true),
    MODALS_POLICY(Flags::enableDesktopWindowingModalsPolicy, true),
    THEMED_APP_HEADERS(Flags::enableThemedAppHeaders, true),
    QUICK_SWITCH(Flags::enableDesktopWindowingQuickSwitch, true),
    APP_HEADER_WITH_TASK_DENSITY(Flags::enableAppHeaderWithTaskDensity, true),
    TASK_STACK_OBSERVER_IN_SHELL(Flags::enableTaskStackObserverInShell, true),
    SIZE_CONSTRAINTS(Flags::enableDesktopWindowingSizeConstraints, true),
    DISABLE_SNAP_RESIZE(Flags::disableNonResizableAppSnapResizing, true),
    DYNAMIC_INITIAL_BOUNDS(Flags::enableWindowingDynamicInitialBounds, false),
    ENABLE_DESKTOP_WINDOWING_TASK_LIMIT(Flags::enableDesktopWindowingTaskLimit, true),
    BACK_NAVIGATION(Flags::enableDesktopWindowingBackNavigation, true),
    EDGE_DRAG_RESIZE(Flags::enableWindowingEdgeDragResize, true),
    TASKBAR_RUNNING_APPS(Flags::enableDesktopWindowingTaskbarRunningApps, true);

    /**
     * Determines state of flag based on the actual flag and desktop mode developer option
     * overrides.
     */
    fun isEnabled(context: Context): Boolean =
        if (!Flags.showDesktopWindowingDevOption() ||
            !shouldOverrideByDevOption ||
            context.contentResolver == null) {
            flagFunction()
        } else {
            val shouldToggleBeEnabledByDefault =
                DesktopModeStatus.shouldDevOptionBeEnabledByDefault()
            when (getToggleOverride(context)) {
                ToggleOverride.OVERRIDE_UNSET -> flagFunction()
                // When toggle override matches its default state, don't override flags. This helps
                // users reset their feature overrides.
                ToggleOverride.OVERRIDE_OFF ->
                    if (shouldToggleBeEnabledByDefault) false else flagFunction()
                ToggleOverride.OVERRIDE_ON ->
                    if (shouldToggleBeEnabledByDefault) flagFunction() else true
            }
        }

    private fun getToggleOverride(context: Context): ToggleOverride {
        val override =
            cachedToggleOverride
                ?: run {
                    val override = getToggleOverrideFromSystem(context)
                    // Cache toggle override the first time we encounter context. Override does not
                    // change with context, as context is just used to fetch Settings.Global
                    cachedToggleOverride = override
                    Log.d(TAG, "Toggle override initialized to: $override")
                    override
                }

        return override
    }

    private fun getToggleOverrideFromSystem(context: Context): ToggleOverride =
        convertToToggleOverrideWithFallback(
            Settings.Global.getInt(
                context.contentResolver,
                Settings.Global.DEVELOPMENT_OVERRIDE_DESKTOP_MODE_FEATURES,
                ToggleOverride.OVERRIDE_UNSET.setting),
            ToggleOverride.OVERRIDE_UNSET)

    /**
     * Override state of desktop mode developer option toggle.
     *
     * @property setting The integer value that is associated with the developer option toggle
     *   override
     */
    enum class ToggleOverride(val setting: Int) {
        /** No override is set. */
        OVERRIDE_UNSET(-1),
        /** Override to off. */
        OVERRIDE_OFF(0),
        /** Override to on. */
        OVERRIDE_ON(1)
    }

    companion object {
        private const val TAG = "DesktopModeFlags"

        /**
         * Local cache for toggle override, which is initialized once on its first access. It needs
         * to be refreshed only on reboots as overridden state is expected to take effect on
         * reboots.
         */
        private var cachedToggleOverride: ToggleOverride? = null

        private val settingToToggleOverrideMap = ToggleOverride.entries.associateBy { it.setting }

        @JvmStatic
        fun convertToToggleOverrideWithFallback(
            overrideInt: Int,
            fallbackOverride: ToggleOverride
        ): ToggleOverride {
            return settingToToggleOverrideMap[overrideInt]
                ?: run {
                    Log.w(TAG, "Unknown toggleOverride int $overrideInt")
                    fallbackOverride
                }
        }
    }
}
